### ISSUE TITLE
Track running stats regardless of track_running_stats=False

### ISSUE THREAD

**Issue by sublee** [pytorch/pytorch#20967]
## 🐛 Bug

Since v1.1.0, `BatchNorm` tracks running stats even its `track_running_stats` option is `False` on the fly.

## To Reproduce

```python
>>> import torch
>>> bn = torch.nn.BatchNorm2d(3)
>>> bn.track_running_stats = False
>>> bn(torch.rand(1, 3, 2, 2))
>>> bn.running_mean
tensor([0.0711, 0.0396, 0.0542])
```

## Expected behavior

It looks like a design issue. If we decide to make `track_running_stats` as an immutable option, to modify this option should raise `AttributeError`.

Or, if we decide to allow to modify `track_running_stats` after initialization, `BatchNorm.forward()` shouldn't track running stats if it is `False`. Here's a possible fix:

```diff
- if self.momentum is None:
+ if self.momentum is None or not self.track_running_stats:
      exponential_average_factor = 0.0
  else:
      exponential_average_factor = self.momentum
```

## Environment

```
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration:
GPU 0: Tesla P40
GPU 1: Tesla P40
GPU 2: Tesla P40
GPU 3: Tesla P40
GPU 4: Tesla P40
GPU 5: Tesla P40
GPU 6: Tesla P40
GPU 7: Tesla P40

Nvidia driver version: 410.104
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.1

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] torch==1.1.0
[pip] torchvision==0.2.1
[conda] Could not collect
```

**Comment by soumith** [comment]
the way it's implemented, `track_running_stats` cant change from False to True (i.e. False in constructor but made True afterwards) because False to True wont create [these buffers](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L35-L37). It can only go from True to False and still be valid.
I think it's better in that case to actually have a way to disable it and enable back which has has some error checking. For example `bn.track_running_stats(False)` -- and internally the function checks that it can only do True -> False -> True and never False -> True by checking for existence of the buffers.

**Comment by sublee** [comment]
As you said it's modification is asymmetric. I think it's a good choice to simply make `track_running_stats` immutable.

**Comment by soumith** [comment]
yea i think that's reasonable

**Comment by ailzhang** [comment]
I think this is fixed by #20985 , thanks for contributing @sublee !

**Comment by jizongFox** [comment]
@sublee  Hi, this issue has not been solved yet (PyTorch=1.3.0). Any alternative ways to deal with BN statistic tracking with `track_running_stats`=False?

**Comment by sublee** [comment]
@jizongFox I didn't fix this issue because my PR #20985 has a backward compatibility issue as @soumith pointed.

**Comment by sublee** [comment]
@ailzhang I guess this issue should be reopened.

**Comment by jizongFox** [comment]
@sublee  I compared the Python code for Batch norm code between 1.0.0 (which works) and recent version, the Python part for forward seemed to be the same. I guess the C/C++ part has some issues then.

**Comment by MaratZakirov** [comment]
May I use bn.eval() & bn.momentum = 0 as workaround?

**Comment by MaratZakirov** [comment]
I had pytorch with version 1.1.0 and consider following simple example of BatchNorm2d
```
import torch
print(torch.__version__)

x = torch.rand(1, 3, 2, 2)

for running_stats in [False, True]:
    for eval in [False, True]:
        print('running_stats:', running_stats, 'eval:', eval)
        bn = torch.nn.BatchNorm2d(3)
        bn.track_running_stats = running_stats
        if eval:
            bn.eval()
        print('\t', bn.running_mean)
        y1 = bn(x)
        print('\t', bn.running_mean)
        y2 = bn(x)
        print('\t', bn.running_mean)
        print('\t', (y1 - y2).sum())
```
The output of which is complete nonsense for me
```
1.1.0
running_stats: False eval: False
	 tensor([0., 0., 0.])
	 tensor([0.0442, 0.0703, 0.0306])
	 tensor([0.0839, 0.1335, 0.0582])
	 tensor(0., grad_fn=<SumBackward0>)
running_stats: False eval: True
	 tensor([0., 0., 0.])
	 tensor([0.0442, 0.0703, 0.0306])
	 tensor([0.0839, 0.1335, 0.0582])
	 tensor(0., grad_fn=<SumBackward0>)
running_stats: True eval: False
	 tensor([0., 0., 0.])
	 tensor([0.0442, 0.0703, 0.0306])
	 tensor([0.0839, 0.1335, 0.0582])
	 tensor(0., grad_fn=<SumBackward0>)
running_stats: True eval: True
	 tensor([0., 0., 0.])
	 tensor([0., 0., 0.])
	 tensor([0., 0., 0.])
	 tensor(0., grad_fn=<SumBackward0>)
```
According to this output running_mean keeps its initial value ONLY in case of model.eval() AND running_stats = True. But the most strange thing is that running_mean has ZERO influence on the output elements difference is always ZERO in all 4 cases. 

Why?!
